backend/services/operation_role.go

```diff
@@ -79,3 +85,63 @@ func SetUserOperationRole(ctx context.Context, db *database.Connection, i SetUse
 	}
 	return nil
 }
+
+func SetUserGroupOperationRole(ctx context.Context, db *database.Connection, i SetUserGroupOperationRoleInput) error {
+	operation, err := lookupOperation(db, i.OperationSlug)
+	if err != nil {
+		return backend.WrapError("Unable to set user group role", backend.UnauthorizedWriteErr(err))
+	}
+
+	if i.UserGroupSlug == "" {
+		return backend.MissingValueErr("User Group Slug")
+	}
+
+	userGroupID, err := userGroupSlugToUserGroupID(db, i.UserGroupSlug)
+	if err != nil {
+		return backend.WrapError("Unable to get user group id from slug", backend.BadInputErr(err, fmt.Sprintf(`No user with slug "%s" was found`, i.UserGroupSlug)))
+	}
+
+	if err := policyRequireWithAdminBypass(ctx, policy.CanModifyUserGroupOfOperation{UserGroupID: userGroupID, OperationID: operation.ID}); err != nil {
+		return backend.WrapError("Unwilling to set user group role", backend.UnauthorizedWriteErr(err))
+	}
+
+	if i.Role == "" {
+		err := db.Delete(sq.Delete("user_group_operation_permissions").Where(sq.Eq{"group_id": userGroupID, "operation_id": operation.ID}))
+
+		if err != nil {
+			return backend.WrapError("Cannot delete user group role", backend.DatabaseErr(err))
+		}
+		return nil
+	}
+
+	var permissions []models.UserGroupOperationPermission
+	err = db.WithTx(context.Background(), func(tx *database.Transactable) {
+		tx.Select(&permissions, sq.Select("*").
+			From("user_group_operation_permissions").
+			Where(sq.Eq{
+				"group_id":     userGroupID,
+				"operation_id": operation.ID,
+			}))
+		if len(permissions) == 0 {
+			tx.Insert("user_group_operation_permissions", map[string]interface{}{
+				"group_id":     userGroupID,
+				"operation_id": operation.ID,
+				"role":         i.Role,
+			})
+		}
+	})
+	if err != nil {
+		return backend.WrapError("Unable to add user role", backend.DatabaseErr(err))
+	}
+
+	if len(permissions) > 0 && permissions[0].Role != i.Role {
+		err = db.Update(sq.Update("user_group_operation_permissions").
+			Set("role", i.Role).
+			Where(sq.Eq{"group_id": userGroupID, "operation_id": operation.ID}))
+
+		if err != nil {
+			return backend.WrapError("Unable to alter user role", backend.DatabaseErr(err))
+		}
```

Comment:
I moved these operations into a transaction liked you asked (which I think was the right call!). Before. I was using a `tx.Get(&permission,...` to get the permissions, and if there was an error (because there was no record), then I'd do the insertion `db.Insert("user_group_operation_permissions", ...`.

However, now that I'm in a transaction I can't rely on the same logic, because if my `Get` doesn't return anything the transaction fails, so I changed `var permission models.UserGroupOperationPermission` to `var permissions []models.UserGroupOperationPermission`, and am checking the length of the slice. This leads to some clunkly logic like `if len(permissions) > 0 && permissions[0].Role != i.Role` instead of `if permission.Role != i.Role {`. 

I've tried to figure out if there's a helper I could create that returns nil if the `Get` is empty, but it seems I would need to modify some code in the Sqlx library, which seems doesn't seem like an option.

Is there a different way you would do this, or does that seem fine?

Reply:
Short version: this is _probably_ the best bet. there are a couple of other routes:
1. Add a clearError function to transactable -- this is handy, but potentially dangerous.
2. Add a special Get function that calls Sqlx.Get under the hood, but checks explicitly for zero row or more than one row, and alerts the caller appropriately -- Still an error, but, one you can check against, and it would explicitly NOT set the transactable error

And I think there are some variations of the above. I'd say it's worth adding that logic if you think you'll need it later, or when you do need it later. For now, this looks okay. You might consider moving the lower if condition (`if len(permissions) > 0` inside the transaction though for additional safety. It would also mean that you don't need to check the length, because it'll be as an `else if` block.

Reply:
> You might consider moving the lower if condition (if len(permissions) > 0 inside the transaction though for additional safety. It would also mean that you don't need to check the length, because it'll be as an else if block.

oh great point! Thanks for the suggestions - yeah I'll leave it that way for now if since that sounds like a decent route